yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts

```diff
@@ -406,37 +402,114 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
+      result.push(
+        ...chunk.map(txHash => ({
+          txHash: TxHash.fromString(txHash),
+          blockHash,
+          isEvictable: !this.#nonEvictableTxs.has(txHash),
+        })),
+      );
+    }
+
+    return result;
+  }
+
+  public async getPendingTxsWithFeePayer(feePayers: AztecAddress[]): Promise<PendingTxInfo[]> {
+    const result: PendingTxInfo[] = [];
+    for (const feePayer of feePayers) {
+      const chunk = await toArray(this.#feePayerToTxHash.getValuesAsync(feePayer.toString()));
+      result.push(...(await Promise.all(chunk.map(txHash => this.getPendingTxInfo(TxHash.fromString(txHash))))));
+    }
+
+    return result;
+  }
+
+  public async getMinedTxHashes(): Promise<[TxHash, BlockNumber][]> {
+    const vals = await toArray(this.#minedTxHashToBlock.entriesAsync());
+    return vals.map(([txHash, blockNumber]) => [TxHash.fromString(txHash), blockNumber]);
+  }
+
+  public async getPendingTxCount(): Promise<number> {
+    return (await this.#pendingTxPriorityToHash.sizeAsync()) ?? 0;
+  }
+
+  public async getMinedTxCount(): Promise<number> {
+    return (await this.#minedTxHashToBlock.sizeAsync()) ?? 0;
+  }
+
+  public async getTxStatus(txHash: TxHash): Promise<'pending' | 'mined' | 'deleted' | undefined> {
+    const key = txHash.toString();
+    const [isMined, isKnown, isDeleted] = await Promise.all([
+      this.#minedTxHashToBlock.hasAsync(key),
+      this.#txs.hasAsync(key),
+      this.#deletedMinedTxHashes.hasAsync(key),
+    ]);
+
+    if (isDeleted) {
+      return 'deleted';
+    } else if (isMined) {
+      return 'mined';
+    } else if (isKnown) {
+      return 'pending';
+    } else {
+      return undefined;
     }
+  }
 
-    if (typeof archivedTxLimit === 'number') {
-      assert(archivedTxLimit >= 0, 'archivedTxLimit must be greater or equal to 0');
-      this.#archivedTxLimit = archivedTxLimit;
+  public updateConfig(cfg: TxPoolOptions): void {
+    if (typeof cfg.archivedTxLimit === 'number') {
+      assert(cfg.archivedTxLimit >= 0, 'archivedTxLimit must be greater or equal to 0');
+      this.#archivedTxLimit = cfg.archivedTxLimit;
     }
 
-    // deletedMinedCleanupThresholdMs is no longer used in block-based cleanup
+    if (this.#evictionManager) {
+      this.#evictionManager.updateConfig(cfg);
+    }
   }
 
   public markTxsAsNonEvictable(txHashes: TxHash[]): Promise<void> {
     txHashes.forEach(txHash => this.#nonEvictableTxs.add(txHash.toString()));
     return Promise.resolve();
   }
 
+  public clearNonEvictableTxs(): Promise<void> {
+    // Clear the non-evictable set after completing the DB updates above.
+    // This ensures pinned (non-evictable) txs are protected while we mark mined txs,
+    // but they won't remain pinned indefinitely across blocks. Note that eviction rules
+    // (including post-mining invalidation) respect the non-evictable flag while it is set.
+    this.#nonEvictableTxs.clear();
+    return Promise.resolve();
+  }
```

Comment:
I know this is not from this PR, but I wonder if we may have a race condition here:

- Our RPC was slow, so we download block N at the same time we get a _proposal_ for block N+1
- The validator client processes the proposal, marking the its txs for block N+1 as non-evictable
- The p2p client processes the downloaded block N, clearing all evictable txs

Maybe the solution is to have a non-evictable flag per block?

Reply:
I agree, this feels like a problem.